com.unity.ml-agents/Runtime/Sensors/Reflection/ObservableAttribute.cs

```diff
@@ -0,0 +1,237 @@
+    /// For each field or property that uses ObservableAttribute, a corresponding
+    /// <see cref="ISensor"/> will be created during Agent initialization, and this
+    /// sensor will read the values during training and inference.
+    /// </summary>
+    /// <remarks>
+    /// ObservableAttribute is intended to make initial setup of an Agent easier. Because it
+    /// uses reflection to read the values of fields and properties at runtime, this may
+    /// be much slower than reading the values directly. If the performance of
+    /// ObservableAttribute is an issue, you can get the same functionality by overriding
+    /// <see cref="Agent.CollectObservations(VectorSensor)"/> or creating a custom
+    /// <see cref="ISensor"/> implementation to read the values without reflection.
+    ///
+    /// Note that you do not need to adjust the VectorObservationSize in
+    /// <see cref="Unity.MLAgents.Policies.BrainParameters"/> when adding ObservableAttribute
+    /// to fields or properties.
+    /// </remarks>
+    /// <example>
+    /// This sample class will produce two observations, one for the m_Health field, and one
+    /// for the HealthPercent property.
+    /// <code>
+    /// using Unity.MLAgents;
+    /// using Unity.MLAgents.Sensors.Reflection;
+    ///
+    /// public class MyAgent : Agent
+    /// {
+    ///     [Observable]
+    ///     int m_Health;
+    ///
+    ///     [Observable]
+    ///     float HealthPercent
+    ///     {
+    ///         get => return 100.0f * m_Health / float(m_MaxHealth);
+    ///     }
+    /// }
+    /// </code>
+    /// </example>
+    [AttributeUsage(AttributeTargets.Field | AttributeTargets.Property)]
+    public class ObservableAttribute : Attribute
+    {
+        string m_Name;
+        int m_NumStackedObservations;
+
+        const BindingFlags k_BindingFlags = BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic;
+        static Dictionary<Type, int> s_TypeSizes = new Dictionary<Type, int>()
+        {
+            {typeof(int), 1},
+            {typeof(bool), 1},
+            {typeof(float), 1},
+            {typeof(Vector2), 2},
+            {typeof(Vector3), 3},
+            {typeof(Vector4), 4},
+            {typeof(Quaternion), 4},
+        };
+
+        /// <summary>
+        /// ObservableAttribute constructor.
+        /// </summary>
+        /// <param name="name">Optional override for the sensor name. Note that all sensors for an Agent
+        /// must have a unique name.</param>
+        /// <param name="numStackedObservations">Number of frames to concatenate observations from.</param>
+        public ObservableAttribute(string name = null, int numStackedObservations = 1)
+        {
+            m_Name = name;
+            m_NumStackedObservations = numStackedObservations;
+        }
+
+        internal static IEnumerable<(FieldInfo, ObservableAttribute)> GetObservableFields(object o, bool declaredOnly)
```

Comment:
Not sure if these would be cleaner as a single `IEnumerable<(MemberInfo, ObservableAttribute)>` that returned both FieldInfos and PropertyInfos. Downside is that we'd have to try to cast before using in the constructor.

Reply:
Might want to do this for caching, but it's OK for now.